Rules/CheckDotNetTypes.cs

```diff
@@ -0,0 +1,176 @@
+                ReflectionTypeAnalysis.Instance.GetTypeFromAssembly(path +"../../Engine/ReferenceAssembly");
+                typeMethod = ReflectionTypeAnalysis.Instance.typeMethod;
+                typeProperty = ReflectionTypeAnalysis.Instance.typeProperties;
+                typeFields = ReflectionTypeAnalysis.Instance.typeFields;
+
+                foreach (MemberExpressionAst member in memberExpressions)
+                {
+                    if (member.Expression is VariableExpressionAst)
+                    {
+                        VariableExpressionAst varAst = member.Expression as VariableExpressionAst;
+
+                        //    //Get the type of the variable before checking if the type has the underliend methods/properties
+                        string type = Helper.Instance.GetVariableTypeFromAnalysis(varAst, func);
+                        string field = member.Member.Extent.Text;
+
+                        if (!typeMethod.ContainsKey(type))
+                        {
+                            if (
+                                !string.IsNullOrEmpty(type) &&
+                                !string.Equals("Microsoft.Windows.Powershell.ScriptAnalyzer.Unreached", type,
+                                    StringComparison.OrdinalIgnoreCase) &&
+                                !string.Equals("Microsoft.Windows.Powershell.ScriptAnalyzer.Undetermined", type,
+                                    StringComparison.OrdinalIgnoreCase) &&
+                                !type.Contains("[]"))
+                            {
+                                yield return
+                                    new DiagnosticRecord(
+                                        string.Format(CultureInfo.CurrentCulture, Strings.DotNetTypeUnavailableError,
+                                            type),
+                                        varAst.Extent, GetName(), DiagnosticSeverity.Warning, fileName);
+                            }
+                        }
+                        else
+                        {
+                            if (!typeMethod[type].Contains(field, StringComparer.OrdinalIgnoreCase) && !typeProperty[type].Contains(field, StringComparer.OrdinalIgnoreCase)
+                                && !typeFields[type].Contains(field,StringComparer.OrdinalIgnoreCase))
+                            {
+                                yield return new DiagnosticRecord(string.Format(CultureInfo.CurrentCulture, Strings.DotNetTypeMemberUnavailableError, type, field),
+                                varAst.Extent, GetName(), DiagnosticSeverity.Warning, fileName);
+                            }
+                        }
+                    }
+
+                    else if (member.Expression is TypeExpressionAst)
+                    {
+                        TypeExpressionAst typeAst = member.Expression as TypeExpressionAst;
+                        Type type = typeAst.TypeName.GetReflectionType();
```

Comment:
Is there a chance that this may be null, i.e for something like []::method()? This is a syntax error but I'm not sure what is the value of type in this case.